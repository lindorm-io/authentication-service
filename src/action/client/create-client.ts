import Joi from "@hapi/joi";
import { IKoaAuthContext } from "../../typing";
import { assertAccountAdmin, encryptClientSecret } from "../../support";
import { Client } from "@lindorm-io/koa-client";

export interface ICreateClientOptions {
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface ICreateClientData {
  clientId: string;
}

const schema = Joi.object({
  description: Joi.string(),
  emailAuthorizationUri: Joi.string().uri(),
  name: Joi.string(),
  secret: Joi.string().allow(null).min(32),
});

export const createClient = (ctx: IKoaAuthContext) => async (
  options: ICreateClientOptions,
): Promise<ICreateClientData> => {
  await schema.validateAsync(options);

  const { account, cache, logger, repository } = ctx;
  const { description, emailAuthorizationUri, name, secret } = options;

  await assertAccountAdmin(ctx)();

  const client = new Client({
    description,
    extra: emailAuthorizationUri ? { emailAuthorizationUri } : {},
    name,
    secret: secret && { signature: encryptClientSecret(secret), updated: new Date() },
  });

  await repository.client.create(client);
  await cache.client.create(client);

  logger.debug("client created", {
    accountId: account.id,
    clientId: client.id,
  });

  return { clientId: client.id };
};
