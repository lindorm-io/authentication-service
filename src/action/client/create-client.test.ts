import MockDate from "mockdate";
import { baseHash } from "@lindorm-io/core";
import { createClient } from "./create-client";
import { getTestAccount, getTestCache, getTestRepository, inMemoryCache, inMemoryStore, logger } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertAccountAdmin: jest.fn(() => () => {}),
  encryptClientSecret: jest.fn((input) => baseHash(input)),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("createClient", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      account: getTestAccount("email@lindorm.io"),
      cache: await getTestCache(),
      logger,
      repository: await getTestRepository(),
    };
  });

  test("should create client", async () => {
    await expect(
      createClient(ctx)({
        description: "description",
        emailAuthorizationUri: "https://lindorm.io/",
        name: "name",
        secret: "45C409CEC3854E3E94500A55865370F5256AD872ED8A46FF8863AD87DB334EF8",
      }),
    ).resolves.toStrictEqual({
      clientId: "be3a62d1-24a0-401c-96dd-3aff95356811",
    });

    expect(inMemoryCache).toMatchSnapshot();
    expect(inMemoryStore).toMatchSnapshot();
  });
});
