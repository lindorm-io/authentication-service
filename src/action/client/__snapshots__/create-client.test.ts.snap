// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createClient should create client 1`] = `
Object {
  "Client::be3a62d1-24a0-401c-96dd-3aff95356811": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"be3a62d1-24a0-401c-96dd-3aff95356811\\",\\"version\\":\\"0\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"events\\":[{\\"name\\":\\"client_created\\",\\"payload\\":{\\"approved\\":\\"false\\",\\"description\\":\\"description\\",\\"extra\\":{\\"emailAuthorizationUri\\":\\"https://lindorm.io/\\"},\\"name\\":\\"name\\",\\"secret\\":{\\"signature\\":\\"NDVDNDA5Q0VDMzg1NEUzRTk0NTAwQTU1ODY1MzcwRjUyNTZBRDg3MkVEOEE0NkZGODg2M0FEODdEQjMzNEVGOA==\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\"},\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\"},\\"date\\":\\"2020-01-01T07:00:00.000Z\\"}],\\"approved\\":\\"false\\",\\"description\\":\\"description\\",\\"extra\\":{\\"emailAuthorizationUri\\":\\"https://lindorm.io/\\"},\\"name\\":\\"name\\",\\"secret\\":{\\"signature\\":\\"NDVDNDA5Q0VDMzg1NEUzRTk0NTAwQTU1ODY1MzcwRjUyNTZBRDg3MkVEOEE0NkZGODg2M0FEODdEQjMzNEVGOA==\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\"}},\\"meta\\":{\\"id\\":\\"string\\",\\"version\\":\\"number\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"events\\":[{\\"name\\":\\"string\\",\\"payload\\":{\\"approved\\":\\"boolean\\",\\"description\\":\\"string\\",\\"extra\\":{\\"emailAuthorizationUri\\":\\"string\\"},\\"name\\":\\"string\\",\\"secret\\":{\\"signature\\":\\"string\\",\\"updated\\":\\"date\\"},\\"created\\":\\"date\\",\\"updated\\":\\"date\\"},\\"date\\":\\"date\\"}],\\"approved\\":\\"boolean\\",\\"description\\":\\"string\\",\\"extra\\":{\\"emailAuthorizationUri\\":\\"string\\"},\\"name\\":\\"string\\",\\"secret\\":{\\"signature\\":\\"string\\",\\"updated\\":\\"date\\"}}}",
    "expiresInSeconds": null,
  },
}
`;

exports[`createClient should create client 2`] = `
Object {
  "authentication": Object {
    "client": Object {
      "data": Array [
        Object {
          "approved": false,
          "created": 2020-01-01T07:00:00.000Z,
          "description": "description",
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "client_created",
              "payload": Object {
                "approved": false,
                "created": 2020-01-01T07:00:00.000Z,
                "description": "description",
                "extra": Object {
                  "emailAuthorizationUri": "https://lindorm.io/",
                },
                "name": "name",
                "secret": Object {
                  "signature": "NDVDNDA5Q0VDMzg1NEUzRTk0NTAwQTU1ODY1MzcwRjUyNTZBRDg3MkVEOEE0NkZGODg2M0FEODdEQjMzNEVGOA==",
                  "updated": 2020-01-01T07:00:00.000Z,
                },
                "updated": 2020-01-01T07:00:00.000Z,
              },
            },
          ],
          "extra": Object {
            "emailAuthorizationUri": "https://lindorm.io/",
          },
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "name": "name",
          "secret": Object {
            "signature": "NDVDNDA5Q0VDMzg1NEUzRTk0NTAwQTU1ODY1MzcwRjUyNTZBRDg3MkVEOEE0NkZGODg2M0FEODdEQjMzNEVGOA==",
            "updated": 2020-01-01T07:00:00.000Z,
          },
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
